sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala

```diff
@@ -1052,11 +1052,11 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
  *    function is order irrelevant
  */
 object EliminateSorts extends Rule[LogicalPlan] {
-  def apply(plan: LogicalPlan): LogicalPlan = plan transform {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
```

Comment:
I remember this comment: https://github.com/apache/spark/pull/21072#discussion_r183392394

Reply:
@allisonwang-db does it have to be `transformUp`?

Reply:
Yes. This case is tricky. Let's consider this example: `df.sortBy('a.asc).sortBy('a.asc).sortBy('a.asc)`
```
Sort(a, false, _)  // A
  Sort(a, false, _)  // B
    Sort(a, false, _)  // C
      ...
```
If we don't use `transformUp`here, the rule will keep both Sort B and C since it will not apply the rule to A's child (B) again. Plan after running this rule will be:
```
Sort(a, false, _) // B
  Sort(a, false, _) // C
     ...
```
And the `Once` strategy's idempotency will be broken since Sort B will be removed if we run this rule again.

Reply:
Would something like this work?
```diff
diff --git a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala
index e41e380539..e17013f7e0 100644
--- a/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala
+++ b/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala
@@ -1054,13 +1054,17 @@ object CombineFilters extends Rule[LogicalPlan] with PredicateHelper {
 object EliminateSorts extends Rule[LogicalPlan] {
   // transformUp is needed here to ensure idempotency of this rule when removing consecutive
   // local sorts.
-  def apply(plan: LogicalPlan): LogicalPlan = plan transformUp {
+  def apply(plan: LogicalPlan): LogicalPlan = plan transformDown {
     case s @ Sort(orders, _, child) if orders.isEmpty || orders.exists(_.child.foldable) =>
       val newOrders = orders.filterNot(_.child.foldable)
       if (newOrders.isEmpty) child else s.copy(order = newOrders)
-    case Sort(orders, false, child) if SortOrder.orderingSatisfies(child.outputOrdering, orders) =>
-      child
-    case s @ Sort(_, _, child) => s.copy(child = recursiveRemoveSort(child))
+    case s @ Sort(orders, _, child) =>
+      val sortsRemoved = recursiveRemoveSort(child)
+      if (!s.global && SortOrder.orderingSatisfies(sortsRemoved.outputOrdering, orders)) {
+        sortsRemoved
+      } else {
+        s.copy(child = sortsRemoved)
+      }
     case j @ Join(originLeft, originRight, _, cond, _) if cond.forall(_.deterministic) =>
       j.copy(left = recursiveRemoveSort(originLeft), right = recursiveRemoveSort(originRight))
     case g @ Aggregate(_, aggs, originChild) if isOrderIrrelevantAggs(aggs) =>
```

Not as elegant, but should avoid the extra iterations from `transformUp`